File: go/types/idempotency.go
package types

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"sync"
	"time"
)

type IdempotencyStore interface {
	Get(key string) (interface{}, error)
	Set(key string, value interface{}, ttlMs int) error
}

type idempotencyEntry struct {
	data      interface{}
	expiresAt time.Time
}

type InMemoryIdempotencyStore struct {
	mu    sync.RWMutex
	cache map[string]*idempotencyEntry
	done  chan struct{}
}

func NewInMemoryIdempotencyStore() *InMemoryIdempotencyStore {
	s := &InMemoryIdempotencyStore{
		cache: make(map[string]*idempotencyEntry),
		done:  make(chan struct{}),
	}
	go s.cleanupLoop()
	return s
}

func (s *InMemoryIdempotencyStore) Get(key string) (interface{}, error) {
	s.mu.RLock()
	entry, ok := s.cache[key]
	s.mu.RUnlock()
	if !ok {
		return nil, nil
	}
	if time.Now().After(entry.expiresAt) {
		s.mu.Lock()
		delete(s.cache, key)
		s.mu.Unlock()
		return nil, nil
	}
	return entry.data, nil
}

func (s *InMemoryIdempotencyStore) Set(key string, value interface{}, ttlMs int) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.cache[key] = &idempotencyEntry{
		data:      value,
		expiresAt: time.Now().Add(time.Duration(ttlMs) * time.Millisecond),
	}
	return nil
}

func (s *InMemoryIdempotencyStore) cleanupLoop() {
	ticker := time.NewTicker(1 * time.Minute)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			s.cleanup()
		case <-s.done:
			return
		}
	}
}

func (s *InMemoryIdempotencyStore) cleanup() {
	now := time.Now()
	s.mu.Lock()
	defer s.mu.Unlock()
	for k, v := range s.cache {
		if now.After(v.expiresAt) {
			delete(s.cache, k)
		}
	}
}

func (s *InMemoryIdempotencyStore) Dispose() {
	close(s.done)
	s.mu.Lock()
	defer s.mu.Unlock()
	s.cache = make(map[string]*idempotencyEntry)
}

func GenerateIdempotencyKey(method, url string, body interface{}) string {
	var bodyBytes []byte
	if body != nil {
		bodyBytes, _ = json.Marshal(body)
	}
	h := sha256.Sum256([]byte(method + ":" + url + ":" + string(bodyBytes)))
	return "mpesa-idem-" + hex.EncodeToString(h[:8])
}

File: typescript/src/utils/idempotency.ts
import { createHash } from "crypto";

export interface IdempotencyStore {
  get(key: string): Promise<unknown | null>;
  set(key: string, value: unknown, ttlMs: number): Promise<void>;
}

export class InMemoryIdempotencyStore implements IdempotencyStore {
  private cache = new Map<string, { data: unknown; expiresAt: number }>();
  private cleanupTimer: ReturnType<typeof setInterval> | null = null;

  constructor(_cleanupIntervalMs = 60_000) {
    this.cleanupTimer = setInterval(() => this.cleanup(), _cleanupIntervalMs);
    this.cleanupTimer.unref();
  }

  async get(key: string): Promise<unknown | null> {
    const entry = this.cache.get(key);
    if (!entry) return null;
    if (Date.now() > entry.expiresAt) {
      this.cache.delete(key);
      return null;
    }
    return entry.data;
  }

  async set(key: string, value: unknown, ttlMs: number): Promise<void> {
    this.cache.set(key, { data: value, expiresAt: Date.now() + ttlMs });
  }

  private cleanup(): void {
    const now = Date.now();
    for (const [key, entry] of this.cache.entries()) {
      if (now > entry.expiresAt) {
        this.cache.delete(key);
      }
    }
  }

  dispose(): void {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
    this.cache.clear();
  }
}

export function generateIdempotencyKey(method: string, url: string, body?: unknown): string {
  const bodyStr = body ? JSON.stringify(body) : "";
  const hash = simpleHash(`${method}:${url}:${bodyStr}`);
  return `mpesa-idem-${hash}`;
}

function simpleHash(str: string): string {
  return createHash("sha256").update(str).digest("hex").slice(0, 16);
}
